src/data/markdown/docs/02 javascript api/10 k6-data/10-SharedArray.md

```diff
@@ -0,0 +1,44 @@
+---
+title: SharedArray
+---
+
+`SharedArray` is an array-like object that shares the underlying memory between VUs. It constructor
+takes a name for the `SharedArray` and a function which needs to return an array object itself. The
+function will be executed only once and it's result will be then be saved in memory once and copies
+of the elements will be given when requested. The name is needed as VUs are completely separate js
+VMs and k6 needs some way to identify the `SharedArray`s that it needs to return. 
+
+This does mean that you can have multiple such ones and even only load some of them for given VUs, although that is
+unlikely to have any performance benefit.
+
+Everything about `SharedArray` is read-only once it is constructed, so it is not possible to
+communicate between VUs using it.
+
+Supported operations include:
+1. getting the number of elements with `length`
+2. getting an element by it's index using the normal syntax `array[index]`
+3. using `for-of` loops
```

Comment:
Let's mention the performance penalty of using `for-of` loops and that a standard C-like loop is preferred.

Reply:
I don't think the performance penalty is interesting the moment you do anything with that data. And I am pretty sure the majority of the penalty comes from the copying of the data, not the actual semantics

That will also change, given that I will be optimizing it using the recently merged changes in goja, on the next update. And also that the goja author might implement something like what we did directly in goja and then we will move to it - either way.

Reply:
OK, but _right now_ there's a performance penalty we should mention, so that people who need it use the plain loop instead.

Reply:
I am still of the opinion that while you might be (look below for why you are not) technically correct - there just isn't a practical reason to not use it, as the performance difference will disappear the moment you use the data for anything. 

I also think you are confusing the difference in the for-of if we don't use SharedArray at all -where there is an extremely big performance difference, but again, that IMO has no practical applications, as you still need to do something with the data.. 

I did quick not very scientific benchmark where the test file is:
```
{"user": "myuser", "password": "password12345"},
```
100k times I ran with 1 VU 10 iterations... and:L

```javascript
    var len  = s.length
    for (var i = 0; i < len; i++) { s[i] }
```
did:
```
     iteration_duration...: avg=2.75s min=2.51s med=2.67s max=3.25s p(90)=3.2s p(95)=3.23s
     iterations...........: 10  0.362695/s
```


```javascript
    for (var i = 0; i < s.length; i++) { s[i] }
```
did:
```
     iteration_duration...: avg=3.29s min=3.2s med=3.29s max=3.38s p(90)=3.35s p(95)=3.36s
     iterations...........: 10  0.303433/s
```

```javascript
    for (var v of s) {}
```
did:
```
     iteration_duration...: avg=2.28s min=1.96s med=2.14s max=2.8s p(90)=2.69s p(95)=2.74s
     iterations...........: 10  0.437979/s
```
So the for-of one is faster (barely ... possibly because `s[i]` taking the array element needs to go through more ... checks. But again IMO the whole thing doesn't matter because this is iterating over 100k elements and doing nothing but just getting them which is not useful. So accessing 1 element takes 28ms, which while not great IMO, is nothing compared to how long it will take to make an HTTP request. 
And again this is not the for-of loop ... this is just the copying of the element (and all the other things that need to be done in order to get the element).

Reply:
I did the for-of loop directly on the underlying array, not on the SharedArray and got
```
     iteration_duration...: avg=106.94ms min=78.2ms med=93.9ms max=161.07ms p(90)=149.06ms p(95)=155.07ms
     iterations...........: 10  4.745281/s
```
so it's more than 10x faster ... again not really useful given that you will need to do something with that data.

Reply:
I guess we can add a note in the vein of "accessing elements has a bigger cost  than with a normal array, but that will balance out in favor of SharedArray for arrays of considerable size, and is likely to be unnoticeable for smaller arrays as well"

Reply:
Hhmm then I was confused by [this comment](https://github.com/dop251/goja/issues/232#issuecomment-745210887) and the discussion in #1739, as my impression was that the wrapping done to implement the iterator would be much less performant, but now I'm still confused as to how it can be faster...

Anyway, if you think none of this should be mentioned I'm OK with it.

Reply:
see https://github.com/loadimpact/k6-docs/pull/196/commits/f90b0b3553138290abd63aaf9f98645f07ad5ee8